core/host_api/impl/child_storage_extension.cpp

```diff
@@ -0,0 +1,231 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "host_api/impl/child_storage_extension.hpp"
+
+#include "runtime/common/runtime_transaction_error.hpp"
+#include "runtime/memory_provider.hpp"
+#include "runtime/ptr_size.hpp"
+#include "runtime/trie_storage_provider.hpp"
+#include "scale/encode_append.hpp"
+
+#include <tuple>
+#include <utility>
+
+using kagome::common::Buffer;
+
+namespace kagome::host_api {
+  using namespace kagome::common::literals;
+
+  ChildStorageExtension::ChildStorageExtension(
+      std::shared_ptr<runtime::TrieStorageProvider> storage_provider,
+      std::shared_ptr<const runtime::MemoryProvider> memory_provider)
+      : storage_provider_(std::move(storage_provider)),
+        memory_provider_(std::move(memory_provider)),
+        logger_{log::createLogger("ChildStorageExtension",
+                                  "child_storage_extension")} {
+    BOOST_ASSERT_MSG(storage_provider_ != nullptr, "storage batch is nullptr");
+    BOOST_ASSERT_MSG(memory_provider_ != nullptr, "memory provider is nullptr");
+  }
+
+  outcome::result<common::Buffer> make_prefixed_child_storage_key(
+      const common::Buffer &child_storage_key) {
+    return ":child_storage:default:"_buf.putBuffer(child_storage_key);
+  }
+
+  template <typename R, typename F, typename... Args>
+  outcome::result<R> ChildStorageExtension::executeOnChildStorage(
+      const common::Buffer &child_storage_key, F func, Args &&... args) const {
+    OUTCOME_TRY(prefixed_child_key,
+                make_prefixed_child_storage_key(child_storage_key));
+    OUTCOME_TRY(child_batch,
+                storage_provider_->getChildBatchAt(prefixed_child_key));
+
+    auto result = func(child_batch, std::forward<Args>(args)...);
+    if (not result) {
+      storage_provider_->clearChildBatches();
+      return result;
+    }
+
+    OUTCOME_TRY(new_child_root, child_batch->commit());
+    OUTCOME_TRY(storage_provider_->getCurrentBatch()->put(
+        prefixed_child_key,
+        common::Buffer{scale::encode(new_child_root).value()}));
+    storage_provider_->clearChildBatches();
+    if constexpr (!std::is_void_v<R>) {
+      return result;
+    }
+    return outcome::success();
+  }
+
+  template <typename... Args>
+  auto ChildStorageExtension::toBuffer(runtime::Memory &memory,
```

Comment:
I think that it does not just convert its arguments to buffers, it loads them from memory, so it'd be better if the function name were descriptive about it.

Reply:
It's currently a tiny local helper, but I renamed it. If we want to use it elsewhere, I'll move it to some utils file.